File: Data Structures & Algorithms/three-integer-sum/submission-0.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        HashMap<Integer, HashSet<Integer>> hMap = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (hMap.containsKey(nums[i])) hMap.get(nums[i]).add(i);
            else {
                HashSet<Integer> hSet = new HashSet<>();
                hSet.add(i);
                hMap.put(nums[i], hSet);
            }
        }
        HashSet<List<Integer>> aList = new HashSet<>();
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                int target = - nums[i] - nums[j];
                HashSet<Integer> h = hMap.get(target);
                if (h != null) {
                    for (int x : h) {
                        if (x > i && x > j) {
                            List<Integer> t = new ArrayList<>();
                            t.add(nums[i]); t.add(nums[j]); t.add(nums[x]);
                            t.sort(null);
                            aList.add(t);
                        }
                    }
                }
            }
        }
        List<List<Integer>> ret = new ArrayList<>();
        for (List<Integer> l : aList) {
            ret.add(l);
        }
        return ret;
    }
}

File: Data Structures & Algorithms/4sum/submission-0.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        ret = set()
        for i in range(len(nums)):
            for j in range(i + 1, len(nums) - 1):
                find = target - nums[i] - nums[j]
                l, r = j + 1, len(nums) - 1
                while l < r:
                    if nums[l] + nums[r] == find:
                        n = [i, j, l, r]
                        l += 1
                        n.sort()
                        if len(n) != 4:
                            continue;
                        ret.add((nums[n[0]], nums[n[1]], nums[n[2]], nums[n[3]]))
                    elif nums[l] + nums[r] < find:
                        l += 1
                    else:
                        r -= 1
        ret2 = [list(i) for i in ret]
        return ret2